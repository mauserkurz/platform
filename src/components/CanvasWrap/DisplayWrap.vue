<template>
  <div class="display-wrap">
    <Display/>
  </div>
</template>

<script>
// TODO add logic to store
// TODO store unit test
// TODO unit test
// TODO add es6 - es10 babel plugins
// TODO add exit game
// TODO add restart level
// TODO add restart game
// TODO add pause
// TODO add health
// TODO add monster
// TODO add help with keys and rules
import { mapState, mapMutations, mapActions } from 'vuex';
import GAME_LEVELS from '@/gameLevels';
import Display from '@/components/Display/Display.vue';
import { ARROW_KEY_LIST, STATUS_MAP } from '@/consts';

// TODO add comments
export default {
  name: 'DisplayWrap',

  components: { Display },

  data() {
    return { arrowKeys: {} };
  },

  computed: {
    ...mapState('core', ['status']),
  },

  methods: {
    ...mapMutations('level', ['ADD_END_LEVEL', 'ADD_END_LEVEL']),
    ...mapActions('core', ['start', 'update']),

    trackArrowKeys() {
      const track = (event) => {
        if (this.status !== STATUS_MAP.PLAYING) {
          this.resetArrowKeys();
          return;
        }

        if (ARROW_KEY_LIST.includes(event.key)) {
          this.arrowKeys[event.key] = event.type === 'keydown';
          event.preventDefault();
        }
      };
      window.addEventListener('keydown', track);
      window.addEventListener('keyup', track);
    },

    resetArrowKeys() {
      this.arrowKeys = {};
    },

    runAnimation(frameFunc) {
      const SECOND = 1000;
      const TICK = 100;
      let lastTime = null;

      const frame = (time) => {
        if (lastTime !== null) {
          const timeStep = Math.min(time - lastTime, TICK) / SECOND;

          if (frameFunc(timeStep) === false) {
            return;
          }
        }
        lastTime = time;
        requestAnimationFrame(frame);
      };
      requestAnimationFrame(frame);
    },

    async runLevel(plan) {
      let ending = 1;

      this.start(plan);
      this.ADD_END_LEVEL();

      return new Promise((resolve) => {
        this.runAnimation((time) => {
          this.update({ time, keys: this.arrowKeys });

          if (this.status === STATUS_MAP.PLAYING) {
            return true;
          }
          if (ending > 0) {
            ending -= time;
            return true;
          }
          this.ADD_END_LEVEL();
          resolve(this.status);
          return false;
        });
      });
    },

    async runGame(plans) {
      this.trackArrowKeys();

      for (let level = 0; level < plans.length;) {
        // eslint-disable-next-line no-await-in-loop
        const status = await this.runLevel(plans[level]);

        if (status === STATUS_MAP.WON) {
          level += 1;
        } else if (status === STATUS_MAP.LOST) {
          // TODO replace alert with component notificator
          // eslint-disable-next-line no-alert
          alert('You lost');
        }
        this.resetArrowKeys();
      }
      // TODO replace alert with component notificator
      // eslint-disable-next-line no-alert
      alert("You've won!");
    },
  },

  mounted() {
    this.runGame(GAME_LEVELS);
  },
};
</script>

<style scoped lang="less">
.display-wrap {}
</style>
