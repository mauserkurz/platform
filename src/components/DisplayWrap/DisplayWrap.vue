<template>
  <div class="display-wrap">
    <v-alert
      v-if="isLost"
      type="error"
      dismissible>You lost</v-alert>
    <v-alert
      v-if="isAllLevelCompleted"
      type="success"
      dismissible>You've won!</v-alert>
    <Display/>
  </div>
</template>

<script>
import { mapState, mapMutations, mapActions } from 'vuex';
import GAME_LEVELS from '@/gameLevels';
import Display from '@/components/Display/Display.vue';
import { ARROW_KEY_LIST, STATUS_MAP } from '@/consts';

export default {
  name: 'DisplayWrap',

  components: { Display },

  data() {
    return {
      arrowKeys: {},
      isAllLevelCompleted: false,
    };
  },

  computed: {
    ...mapState('core', ['status']),

    isWon() {
      return this.status === STATUS_MAP.WON;
    },

    isLost() {
      return this.status === STATUS_MAP.LOST;
    },

    isPlaying() {
      return this.status === STATUS_MAP.PLAYING;
    },
  },

  methods: {
    ...mapMutations('level', ['ADD_END_LEVEL']),
    ...mapActions('core', ['update']),
    ...mapActions('level', ['start']),

    /**
     * Начнет отслеживать события нажатия клавишь, указанных в ARROW_KEY_LIST
     * записывая зажатые клавиши в arrowKeys
     */
    trackArrowKeys() {
      const track = (event) => {
        if (!this.isPlaying) {
          this.resetArrowKeys();
          return;
        }

        if (ARROW_KEY_LIST.includes(event.key)) {
          this.arrowKeys[event.key] = event.type === 'keydown';
          event.preventDefault();
        }
      };
      window.addEventListener('keydown', track);
      window.addEventListener('keyup', track);
    },

    /**
     * Сбросит записи о нажатых клавишах, чтобы можно было останавливать
     * игрока при смене статуса игры
     */
    resetArrowKeys() {
      this.arrowKeys = {};
    },

    /**
     * Запустит анимацию, выполняя в каждом кадре переданную функцию
     * @param {function} frameFunc - коллбэк, выполняющий обновление данных уровня
     */
    runAnimation(frameFunc) {
      const SECOND = 1000;
      const TICK = 100;
      let lastTime = null;

      const frame = (time) => {
        if (lastTime !== null) {
          const timeStep = Math.min(time - lastTime, TICK) / SECOND;

          if (!frameFunc(timeStep)) {
            return;
          }
        }
        lastTime = time;
        requestAnimationFrame(frame);
      };
      requestAnimationFrame(frame);
    },

    /**
     * Подготовит дынные об уровне, используя переданный шаблон, запустит анимацию, во время которой
     * будут сменяться данные о уровне и перересовываться canvas
     * @param {string} plan - строка шаблон уровня
     * @return {Promise<undefined>}
     */
    async runLevel(plan) {
      let ending = 1;

      this.start(plan);

      return new Promise((resolve) => {
        this.runAnimation((time) => {
          this.update({ time, keys: this.arrowKeys });

          if (this.isPlaying) {
            return true;
          }
          if (ending > 0) {
            ending -= time;
            return true;
          }
          this.ADD_END_LEVEL();
          resolve();
          return false;
        });
      });
    },
  },

  /**
   * Начнет отслеживание нажатых клавишь и стартует последовательный запуск уровеней
   * при прохождении или перезапуск при проигрыше
   * @return {Promise<void>}
   */
  async mounted() {
    this.trackArrowKeys();

    for (let level = 0; level < GAME_LEVELS.length;) {
      // eslint-disable-next-line no-await-in-loop
      await this.runLevel(GAME_LEVELS[level]);

      if (this.isWon) {
        level += 1;
      }
      this.resetArrowKeys();
    }
    this.isAllLevelCompleted = true;
  },
};
</script>

<style scoped lang="less">
.display-wrap {}
</style>
