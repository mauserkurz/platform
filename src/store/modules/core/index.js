import { setProp } from '@/store/reusable';
import { PLAYER_CONFIG, LAVA_CONFIG, STATUS_MAP } from '@/consts';

/**
 * Вернет значение отвечающее на вопрос пересекает ли сущность actor1 другую сущность actor2
 * @param {object} actor1 - сущность, пересечение которой мы ищем
 * @param {object} actor2 - сущность с которой может пересечься actor1
 * @return {boolean} - значение пересечения
 */
const overlap = (actor1, actor2) => actor1.pos.x + actor1.size.x > actor2.pos.x
  && actor1.pos.x < actor2.pos.x + actor2.size.x
  && actor1.pos.y + actor1.size.y > actor2.pos.y
  && actor1.pos.y < actor2.pos.y + actor2.size.y;

export default {
  namespaced: true,

  state: {
    actors: [],
    status: '',
  },

  getters: {
    /**
     * Найдет игрока среди подвижных сущностей на текущем уровне
     * @param {object} state - состояние стора
     * @return {object} - сущность игрока
     */
    player(state) {
      return state.actors.find(a => a.type === PLAYER_CONFIG.TYPE);
    },

    /**
     * Вернет данные текущего уровня, изменения которых можно отслеживать для
     * @param {array} actors - подвижные объекты на уровне
     * @param {string} status - статус уровня, который может быть - победа, поражение или играет
     * последующей перерисовки во view
     * @return {array} - данные текущего уровня
     */
    stateChange({ actors, status }) {
      return [actors, status];
    },
  },

  mutations: {
    /**
     * Сохранит список подвижных сущностей на уровне
     */
    SET_ACTORS: setProp('actors'),

    /**
     * Сохранит статус игры
     */
    SET_STATUS: setProp('status'),
  },

  actions: {
    /**
     * Обновит состояние уровня в завистимости от действий подвижных сущностей
     * и действий игрока
     * @param {object} context - содержит данные стора и функции для управления им
     * @param {number} time - время в секундах, показываеющее разницу между прошлым
     * и следующим кадрами анимации
     * @param {object} keys - map в котором указаны нажатые клавиши
     * @return {string} - статус игры после обновления состояния уровня
     */
    update(context, { time, keys }) {
      const {
        state,
        getters,
        rootGetters,
        commit,
      } = context;
      const touches = rootGetters['level/touches'];
      const actors = state.actors
        .map(actor => actor.update(time, touches, keys));

      commit('SET_ACTORS', actors);
      if (state.status !== STATUS_MAP.PLAYING) {
        return state.status;
      }
      const { player } = getters;

      if (touches(player.pos, player.size, LAVA_CONFIG.TYPE)) {
        commit('SET_STATUS', STATUS_MAP.LOST);
        return state.status;
      }

      for (const actor of actors) {
        if (actor !== player && overlap(actor, player)) {
          const newState = actor.collide({ actors, status: state.status });

          commit('SET_ACTORS', newState.actors);
          commit('SET_STATUS', newState.status);
        }
      }
      return state.status;
    },
  },
};
