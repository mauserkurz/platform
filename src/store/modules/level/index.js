import { setProp, plus } from '@/store/reusable';
import createVector from '@/models/vector/';
import createPlayer from '@/models/player/';
import createLava from '@/models/lava/';
import createInitialCoin from '@/models/coin/';
import {
  LAVA_CONFIG,
  WALL_CONFIG,
  EMPTY_CONFIG,
  STATUS_MAP,
} from '@/consts';

const LEVEL_CHAR_MAP = {
  '.': EMPTY_CONFIG.TYPE,
  '#': WALL_CONFIG.TYPE,
  '+': LAVA_CONFIG.TYPE,
  '=': createLava,
  '|': createLava,
  v: createLava,
  o: createInitialCoin,
  '@': createPlayer,
};

export default {
  namespaced: true,

  state: {
    levelHeight: 0,
    levelWidth: 0,
    countEndLevel: 0,
    rows: [[]],
  },

  getters: {
    /**
     * Создаст функцию, чтобы можно было проверить пересечение сущности с краем
     * уровня или другой сущностью переданного типа (type), применяется для обработки
     * движения и с учетом столкновения с краем уровня или стеной
     * @param {object} state - состояние стора
     * @return {function} - функция для проверки пересечения, где аргументы
     * pos - позиция сущности,
     * size - размер сущности
     * type - тип сущности, пересечение с которой требуется найти
     * вернет булево значение, при этом true - когда сущность будет пересекать
     * с краем уровня или другой сущностью переданного типа (type)
     */
    touches: state => (pos, size, type) => {
      const xStart = Math.floor(pos.x);
      const xEnd = Math.ceil(pos.x + size.x);
      const yStart = Math.floor(pos.y);
      const yEnd = Math.ceil(pos.y + size.y);

      for (let y = yStart; y < yEnd; y += 1) {
        for (let x = xStart; x < xEnd; x += 1) {
          const isOutside = x < 0 || x >= state.levelWidth
            || y < 0 || y >= state.levelHeight;
          const here = isOutside ? WALL_CONFIG.TYPE : state.rows[y][x];

          if (here === type) {
            return true;
          }
        }
      }
      return false;
    },
  },

  mutations: {
    /**
     * Сохранит высоту уровня
     */
    SET_LEVEL_HEIGHT: setProp('levelHeight'),

    /**
     * Сохранит ширину уровня
     */
    SET_LEVEL_WIDTH: setProp('levelWidth'),

    /**
     * Сохранит двумерный массив неподвижных сущностей на уровне
     */
    SET_ROWS: setProp('rows'),

    /**
     * Добавит 1 к счетчику начала уровня, нужно для запуска уровня в компоненте
     * при смене значения счетчика
     */
    ADD_END_LEVEL: plus('countEndLevel'),
  },

  actions: {
    /**
     * Сгенерирует данные уровня на основе переданного шаблона (plan) и сохранит их в state,
     * сгенерированы и сохранены будут ширина и высота уровня,
     * данные сущностей на уровне и статус игры
     * @param {function} commit - запустит мутацию с перереданными параметрами
     * @param {string} plan
     */
    start({ commit }, plan) {
      const startActors = [];
      const rows = plan
        .trim()
        .split('\n')
        .map(row => row.trim().split(''))
        .map((row, y) => row.map((char, x) => {
          const entity = LEVEL_CHAR_MAP[char];

          if (typeof entity === 'string') {
            return entity;
          }

          startActors.push(entity(createVector(x, y), char));
          return EMPTY_CONFIG.TYPE;
        }));

      commit('SET_LEVEL_HEIGHT', rows.length);
      commit('SET_LEVEL_WIDTH', rows[0].length);
      commit('SET_ROWS', rows);
      commit('core/SET_ACTORS', startActors, { root: true });
      commit('core/SET_STATUS', STATUS_MAP.PLAYING, { root: true });
    },
  },
};
