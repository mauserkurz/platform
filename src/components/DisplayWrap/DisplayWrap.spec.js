import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import clone from 'ramda/src/clone';
import core from '@/store/modules/core/';
import level from '@/store/modules/level/';
import GAME_LEVELS from '@/gameLevels';
import DisplayWrap from './DisplayWrap.vue';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Vuetify);

describe('component DisplayWrap', () => {
  const createWrapper = ({ options = {}, modules = {}, needsRendering = false } = {}) => {
    const store = new Vuex.Store({
      namespaced: true,
      modules: Object.assign({}, { core, level }, modules),
    });
    const defaultOptions = { localVue, store, stubs: { Display: '<div class="display"></div>' } };
    const customMount = needsRendering ? mount : shallowMount;
    const wrapper = customMount(DisplayWrap, Object.assign({}, defaultOptions, options));

    return { store, wrapper };
  };

  describe('rendering', () => {
    it('should render html', () => {
      const { wrapper } = createWrapper();

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('mounted', () => {
    it('должен запустить игру через экшн start', () => {
      const levelClone = clone(level);
      const spy = jest.spyOn(levelClone.actions, 'start');

      createWrapper({ modules: { level: levelClone } });
      expect(spy.mock.calls[0][1]).toBe(GAME_LEVELS[0]);
    });

    it('должен по событию нажатия кнопки обновить уровень вызвав экнш update', () => {
      const coreClone = clone(core);
      const spy = jest.spyOn(coreClone.actions, 'update');
      const map = {};
      let frameCallBack = null;

      window.addEventListener = jest.fn((event, cb) => {
        map[event] = cb;
      });
      window.requestAnimationFrame = (frame) => {
        frameCallBack = frame;
      };
      createWrapper({ modules: { core: coreClone } });
      map.keydown({ key: 'ArrowLeft', type: 'keydown', preventDefault() {} });
      frameCallBack(0);
      frameCallBack(50);

      expect(spy.mock.calls[0][1]).toEqual({ keys: { ArrowLeft: true }, time: 0.05 });
    });

    it('должен если статус игры не "играет" по событию нажатия кнопки не реагировать', () => {
      const coreClone = clone(core);
      const spy = jest.spyOn(coreClone.actions, 'update');
      const map = {};
      let frameCallBack = null;

      window.addEventListener = jest.fn((event, cb) => {
        map[event] = cb;
      });
      window.requestAnimationFrame = (frame) => {
        frameCallBack = frame;
      };
      const { store } = createWrapper({ modules: { core: coreClone } });

      store.commit('core/SET_STATUS', 'won');
      map.keydown({ key: 'ArrowUp', type: 'keydown', preventDefault() {} });
      frameCallBack(0);
      frameCallBack(100);

      expect(spy.mock.calls[0][1]).toEqual({ keys: {}, time: 0.1 });
    });

    it(`должен запустить этот же уровень через 1 секунду после поражения через
      экшн start`, async () => {
      const levelClone = clone(level);
      const spy = jest.spyOn(levelClone.actions, 'start');
      let frameCallBack = null;

      window.requestAnimationFrame = (frame) => {
        frameCallBack = frame;
      };
      const { store } = createWrapper({ modules: { level: levelClone } });
      store.commit('core/SET_STATUS', 'lost');
      Array.from({ length: 13 }, (_, index) => frameCallBack(index * 100));

      await flushPromises();
      expect(spy.mock.calls[1][1]).toBe(GAME_LEVELS[0]);
    });

    it(`должен запустить этот новый уровень через 1 секунду после победы через
      экшн start`, async () => {
      const levelClone = clone(level);
      const spy = jest.spyOn(levelClone.actions, 'start');
      let frameCallBack = null;

      window.requestAnimationFrame = (frame) => {
        frameCallBack = frame;
      };
      const { store } = createWrapper({ modules: { level: levelClone } });
      store.commit('core/SET_STATUS', 'won');
      Array.from({ length: 13 }, (_, index) => frameCallBack(index * 100));

      await flushPromises();
      expect(spy.mock.calls[1][1]).toBe(GAME_LEVELS[1]);
    });

    it('должен при звершении всех уровней отобразить сообщений о победе', async () => {
      const levelClone = clone(level);
      let frameCallBack = null;

      window.requestAnimationFrame = (frame) => {
        frameCallBack = frame;
      };
      const { wrapper, store } = createWrapper({ modules: { level: levelClone } });

      for (let i = 0; i < GAME_LEVELS.length; i += 1) {
        store.commit('core/SET_STATUS', 'won');
        // eslint-disable-next-line no-loop-func
        Array.from({ length: 13 }, (_, index) => frameCallBack(index * 100));
        // eslint-disable-next-line no-await-in-loop
        await flushPromises();
      }
      expect(wrapper.find('v-alert-stub').text()).toBe("You've won!");
    });
  });
});
