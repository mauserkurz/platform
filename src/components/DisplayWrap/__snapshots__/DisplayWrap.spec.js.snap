// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component DisplayWrap rendering should render html 1`] = `
<div class="display-wrap">
  <!---->
  <!---->
  <div class="display"></div>
</div>
`;
